### PR TITLE
feat: add precompute partition writer

### PR DESCRIPTION
**Pull request by ZENOTME**
The precompute partition writer accepts the data chunk with a partition value column computed before.

### CODE REVIEW

File: icelake/src/io_v2/functional_writer/precompute_partition_writer.rs [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+            schema: new_schema_builder.finish().into(),
+            partition_type: self.partition_type,
+            row_converter,
+            partition_col_idx: self.partition_col_idx,
+        })
+    }
+}
+
+/// Partition append only writer
+pub struct PrecomputedPartitionedWriter<B: IcebergWriterBuilder> {
```
Comment (liurenjie1024) [inline icelake/src/io_v2/functional_writer/precompute_partition_writer.rs:127]: Please add comments here to explain the motivation behind this writer, and what's more important, the expected layout of inputs.

File: icelake/src/io_v2/functional_writer/precompute_partition_writer.rs [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+    inner_writers: HashMap<OwnedRow, (B::R, StructValue)>,
```
Comment (liurenjie1024) [inline icelake/src/io_v2/functional_writer/precompute_partition_writer.rs:128]: Why we need a `StructValue`?

File: icelake/src/io_v2/functional_writer/precompute_partition_writer.rs [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+            .row_converter
+            .convert_columns(&[partition_value.clone()])
+            .map_err(|e| Error::new(ErrorKind::ArrowError, e.to_string()))?;
+        let mut filters = HashMap::new();
+        rows.into_iter().enumerate().for_each(|(row_id, row)| {
+            filters
+                .entry(row.owned())
+                .or_insert_with(|| (vec![false; batch.num_rows()], row_id))
+                .0[row_id] = true;
+        });
```
Comment (liurenjie1024) [inline icelake/src/io_v2/functional_writer/precompute_partition_writer.rs:165]: This is kind of inefficient, how about generating an array of `(partition_value, row_id)` and then sort them?

File: icelake/src/io_v2/functional_writer/precompute_partition_writer.rs (hunk 2 above)
Comment (liurenjie1024) [reply icelake/src/io_v2/functional_writer/precompute_partition_writer.rs:128]: Oh, I see, please add some comment here.

**Review by liurenjie1024 — COMMENTED** [review]
Generally LGTM, thanks!

File: icelake/src/io_v2/functional_writer/precompute_partition_writer.rs (hunk 3 above)
Comment (ZENOTME) [reply icelake/src/io_v2/functional_writer/precompute_partition_writer.rs:165]: Good point!

File: icelake/src/io_v2/functional_writer/precompute_partition_writer.rs [hunk 4]
```diff
@@ -0,0 +1,347 @@
+// Copyright 2024 RisingWave Labs
```
[the remark this answers is not in the document]
Comment (ZENOTME) [reply icelake/src/io_v2/functional_writer/precompute_partition_writer.rs]: my bad🥵

**Review by liurenjie1024 — APPROVED** [review]
LGTM, thanks!

**Pull request merged**